### ISSUE TITLE
Allow directly calling an object field of a `function` type

### ISSUE THREAD

**Issue by grainier** [ballerina-platform/ballerina-lang#22654]
**Description:**
Following snippet result in `undefined function 'func' in object 'xxxxx/main:0.1.0:_Bar'`, even if it's there.

```ballerina
type _Foo abstract object {
    public function(record{}|error? frame) returns record{}|error? func;
    public function process() returns record{}|error?;
};

type _Bar object {
    *_Foo;

    public function __init(function(record{}|error? frame) returns record{}|error? func) {
        self.func = func;
    }

    public function process() returns record{}|error? {
        record{}|error? cFrame = self.func({});
        return cFrame;
    }
};
```
However, it'll work if we assigning it into a local var and use it like;
```ballerina
    public function process() returns record{}|error? {
        function(record{}|error? frame) returns record{}|error? func = self.func;
        record{}|error? cFrame = func({});
        return cFrame;
    }
```

**Affected Versions:**
1.2.1

**OS, DB, other environment details and versions:**
n/a

**Comment by MaryamZi** [comment]
This seems to be a generic issue when attempting to call a function which is a field of an object.

```ballerina
import ballerina/io;

class Foo {

    function () bar;

    function init(function () bar) {
        self.bar = bar;
    }

    function callBar() {
        self.bar(); // error: undefined method 'bar' in object 'Foo'
    }
}

public function main() {
    Foo f = new (() => io:println("Hello"));

    f.bar(); // error: undefined method 'bar' in object 'Foo'
}
```

IINM, this was not allowed previously, and we had to assign the field's value to a function variable and then call that function. But this should now be allowed according to the spec (see https://github.com/ballerina-platform/ballerina-spec/issues/257).

The method-call-expr is now defined as

> A method-call-expr either calls an object's method, calls an object's field where the type of the field is a function or calls a function in the lang library.

**Comment by MaryamZi** [comment]
Probably fixed in https://github.com/ballerina-platform/ballerina-lang/pull/31796. Looping in @rdhananjaya.

**Comment by rdhananjaya** [comment]
This is fixed with #31796
